### ISSUE TITLE
Any plan for aarch64?

### ISSUE THREAD

**Issue by madper** [Vanilla-OS/iso-builder#87]
### Feature Description

Supports arm64 devices. 
1. As most arm64 devices has no acpi yet  we may have to modify the build process to allow guys specify the dtb/kernel file. 
2. The ABroot implementation for uboot. (I haven't check the current implementation for x86. I bet it's grub logic with grub-env. And I do believe supportting aarch64 needs porting the logic to uboot. )
3. The installer should handle arm devices as well.

### Rationale

Both transcational upgrading and Immutable are great features for edge project. I was going to implement it with ubuntu core. However I'm not a big fan of snap package and I believe Vanilla OS is a far more elegante choice. 

### Related Issues

Vanilla-OS/live-iso#53

**Comment by matthiasjg** [comment]
Totally agree. :raised_hands: 

With the [just started initial work on Vanilla OS 2.0 Orchid](https://vanillaos.org/2023/03/07/vanilla-os-2.0-orchid-initial-work.html) and moving to Debian Sid, any ARM based device—e.g. laptops like the Lenovo ThinkPad X13s Gen 1 w/ it's Qualcomm® Snapdragon™ 8cx Gen 3 SC8280XP prozessor—would be a LOVELY target device.

Especially as there is already [a Debian based installation approach](https://forums.lenovo.com/t5/Other-Linux-Discussions/Does-anybody-know-if-there-is-work-being-done-to-integrate-X13s-ARM-processor-with-linux/m-p/5175315?page=1#5771660). Both, installation approach and hardware support can be expected to improve w/ kernel 6.2 and 6.3 and the current work  w/ in Linux distributions communities (Ubuntu, Fedora, Arch). 

Totally would love run Vanilla on my X13s… and help along the way to make this "a thing". :star_struck:

**Comment by dajix350** [comment]
Who knows, maybe this could lead to VanillaPi

**Comment by iMonZ** [comment]
> Who knows, maybe this could lead to VanillaPi

Please hell no!
What about so many other aarch64 devices?
NAS Devices or MacBooks or other Aarch64 laptops.
Please don't ever release raspberrypi based os!
Thanks!

**Comment by gradygambrell** [comment]
I do want to have an ARM build for Parallels Desktop on M Series Mac. This would be fantastic.

**Comment by ChristinWhite** [comment]
That would be my use case too @gradygambrell, it would be amazing to be able to have access to Vanilla on primary machine.

**Comment by itjustcrashed** [comment]
Vanilla OS Asahi would be amazing. Based hardware with the best of the best software.


If anyone wants to work on this,  I'm happy to help. I'm equipped with an Apple Silicon Mac, and lots of knowledge on Asahi.

**Comment by user858753257** [comment]
Hopefully Orchid have an ARM compatible build 🫣

**Comment by NN708** [comment]
Hello everyone,

I'm excited to share an early test build of Vanilla OS 2 with experimental ARM64 support! I've been working on this for the past few months.

This build is based on Debian Sid with the latest Linux kernel 6.16, so it should be compatible with most ARM64 hardware that has mainline Linux support.

I need your help to test it on as many devices and virtual machines as possible! Please remember this is an early, unsupported test build and is **NOT** ready for production use.

## How to Test

### Prerequisites:

You will need a UEFI-compatible firmware to boot the ISO.

+ **Real Hardware**: Use an EDK2 or UEFI on U-Boot firmware for your hardware.
    + **Raspberry Pi 4**: Flash the [pftf/RPi4](https://github.com/pftf/RPi4) firmware.
    + **RK3588 SBCs** (Orange Pi 5, Rock 5B, etc.): Flash the [edk2-porting/edk2-rk3588](https://github.com/edk2-porting/edk2-rk3588) firmware.
+ **Virtual Machines**: Most modern VMs support UEFI. For Apple Silicon Macs, you can use [UTM](https://mac.getutm.app/).

### Download and Flash:

+ Download the ARM64 test ISO from GitHub Actions: https://github.com/vanilla-arm/live-iso/actions/runs/19174487638
+ To boot on real hardware, flash the downloaded ISO to a USB drive or other media using your preferred tool (e.g., [balenaEtcher](https://etcher.balena.io/)).

### Boot and Install:

+ Boot your device or VM from the installation media.
+ If the Vanilla installer fails to launch, it could be because of a known issue: https://github.com/Vanilla-OS/vanilla-installer/issues/461. Open a terminal and run:
    ```bash
    IGNORE_CPU=1 vanilla-installer
    ```
+ In the installer, choose **"Install Custom Image (Advanced)"**.
    <img src="https://github.com/user-attachments/assets/83de095a-f6ba-44de-b969-3ec0bda1c2c5" />
+ After configuring your locale and network, you will be prompted for a custom image. Enter:
    ```
    ghcr.io/vanilla-arm/desktop:test
    ```
    <img src="https://github.com/user-attachments/assets/ea88bef6-b579-4d12-aa14-a49f35f693b1" />
+ Proceed with the installation.

Once installed, please explore the system and test anything you'd like. Your feedback is invaluable for helping improve Vanilla OS on ARM64!

**Comment by justsaft** [comment]
Testing this in a VM under VMware Fusion (macOS) on an M4 right now.
Runs really well actually.
Installation was mostly fine.

Here's the things that I found not working:

- when booting the iso...
   - there's a really long wait on a black screen (bit shorter with safe graphics but still there)
   - GNOME can't find an "installer" session
<img width="403" height="305" alt="Image" src="https://github.com/user-attachments/assets/a8f1d8d7-5dfb-4c07-bde0-9737bcf8c9ec" />

- the pico container failed to setup during first setup
- pico fails to start when entering
<img width="623" height="397" alt="Image" src="https://github.com/user-attachments/assets/62522a3b-a902-4f40-bad5-2155a8d480cd" />

- hardware acceleration seems to work fine (GNOME runs smoothly)
  - however video playback is broken
I'm not sure who's issue that is (Fusion or Apple). I'm likely missing VM tools aswell however it doesn't run like I'm missing them. It is likely using software decoding

Some VM settings:
- CPU: 4 cores
- RAM: 4GB assigned
- VGA: 3D accel enabled. 256MB VRAM

**Comment by NN708** [comment]
Hi @justsaft, thank you for your detailed feedback!

> * there's a really long wait on a black screen (bit shorter with safe graphics but still there)
> * GNOME can't find an "installer" session

This is a known issue (#397) and is not specific to ARM.

> * the pico container failed to setup during first setup
> * pico fails to start when entering

This is expected at the moment, as the [ghcr.io/vanilla-os/vso](https://ghcr.io/vanilla-os/vso) image isn't yet ready for multi-architecture support. You can try Apx with another OS image that has multi-architecture support to verify if it is working.

I'm glad to hear that general GNOME performance is working smoothly! If you have any further insights or run into additional issues, please let me know.

**Comment by justsaft** [comment]
Hey @NN708, thanks for the efforts on bringing Vanilla to ARM!

I got some more feedback.
  
> > * there's a really long wait on a black screen (bit shorter with safe graphics but still there)
> > * GNOME can't find an "installer" session
> 
> This is a known issue ([#397](https://github.com/Vanilla-OS/live-iso/issues/397)) and is not specific to ARM.

Interesting, I had never seen this on either of my x86_64 machines I installed Vanilla on. Might be because that was a really long time ago.

On a different note, I've also tested UTM now. Installed just about the same as on Fusion, with vso failing to setup as per arch issue as expected.

Re the video playback issue:
- (On VMware Fusion) I tried video playback again on the recommended 8GB VRAM: didn't make a difference.
- I installed VOS again in UTM and there it works. Likely missing `open-vm-tools` on Fusion

Issue with abroot upgrades:
- I seem to be unable to upgrade at all.
- First the default image was configured to be `vanilla-os/desktop` instead of the ARM image. -> I changed it using `pkexec abroot config-editor` to `vanilla-arm/desktop` with tag `test`, from host-shell.
- Now upgrading fails with `Manifest does not contain layer property`
<img width="751" height="200" alt="Image" src="https://github.com/user-attachments/assets/5ab105a8-035b-448a-aa2d-c53eba93a1e2" />

-> entirely prevents me from upgrading and applying packages.
This seems to be specific to the vanilla-arm/desktop:test image, since it _can_ pull regular desktop, this just won't result in a usable system since it's built for amd64.


VM specs:
- virt-10.0
- aarch64
- CPU: default, 5 cores
- GPU: virtio-gpu-gl-pci
- net: virtio-net-pci
- audio: intel-hda (ich6)
    - side note: has a bug where it defaults to 0% volume on each boot
- Use Hypervisor: On

**Comment by NN708** [comment]
@justsaft Thanks for the report! This is definitely an issue with ABRoot.

To help us investigate, could you please run `pkexec abroot upgrade --verbose` and share the full output by opening an issue in [ABRoot](https://github.com/Vanilla-OS/ABRoot)? I'll look into fixing it later.

**Comment by migtorr** [comment]
I'm aware the aarch64 build is for Raspberries but so you know I tried to boot it on the Pinebook Pro and did load but no touchpad gestures, wifi or sound

**Comment by NN708** [comment]
Hi @migtorr, glad to hear you were able to boot it! Just to clarify, this build is designed for UEFI-compatible ARM64 devices in general, not specifically for Raspberry Pi.

Regarding the hardware issues you mentioned (touchpad, Wi-Fi and sound), these depend mainly on two factors:
1. Bootloader compatibility
2. Linux kernel driver support

Since Vanilla OS uses the upstream kernel, hardware support isn't something we directly control. The good news is that support for these features should improve over time as the kernel is updated, though you can expect some delay from upstream.

In the meantime, you may want to check the [kernel mailing lists](https://lore.kernel.org) and your bootloader's documentation for more specific details on why certain hardware isn't working or how to improve support.

**Comment by NN708** [comment]
Hi everyone,

Thank you all for testing the ARM64 build!

I'm pleased to share that most issues with the installer ISO and desktop image have been resolved. We are now building the latest images using the [Vanilla OS testing repository](https://testing.vanillaos.org), featuring Linux kernel 6.17 and GNOME 49.

If you have time, please help test the new build:

+ The latest ISO is available via GitHub Actions: https://github.com/vanilla-arm/live-iso/actions/runs/19174487638
+ The custom desktop image remains `ghcr.io/vanilla-arm/desktop:test`
+ To test the VSO image, edit `/usr/share/vso/apx/stacks/vso-pico.yaml` to:
    ```yaml
    name: vso-pico
    base: ghcr.io/vanilla-arm/vso:test
    packages: []
    pkgmanager: apt
    builtin: true
    ```

Please note that this is still a test build and **NOT** ready for production use.

**Comment by NN708** [comment]
Hi everyone,

A new image is now available for testing, featuring a major change in filesystem structure (https://github.com/Vanilla-OS/ABRoot/issues/463) and an upgrade to our [new package repository](https://repo3.vanillaos.org).

If you have time to help with testing, here are the resources and instructions:

+ The latest ISO is available via GitHub Actions: https://github.com/vanilla-arm/live-iso/actions/runs/20392760963
+ The custom desktop image remains `ghcr.io/vanilla-arm/desktop:test`
+ To test the VSO image, editing `/usr/share/vso/apx/stacks/vso-pico.yaml` is still required. Since `/` is now mounted as read-only, remount it first with:
    ```console
    pkexec mount -o remount,rw /
    ```

Please note that this is still a test build and **NOT** ready for production use.

**Comment by NN708** [comment]
Hi everyone,

Most of the work for ARM64 support is complete, and the changes for the desktop image have now been merged into `dev` branch.

Testing it is now quite simple:

+ The latest ISO is available via GitHub Actions: https://github.com/vanilla-arm/live-iso/actions/runs/20392760963
+ The custom desktop image is now `ghcr.io/vanilla-os/desktop:dev`

Please note that this is still a test build and **NOT** ready for production use.

**Comment by muhammadbahaa2001** [comment]
> Hi everyone,
> 
> Most of the work for ARM64 support is complete, and the changes for the desktop image have now been merged into `dev` branch.
> 
> Testing it is now quite simple:
> 
> * The latest ISO is available via GitHub Actions: https://github.com/vanilla-arm/live-iso/actions/runs/20392760963
> * The custom desktop image is now `ghcr.io/vanilla-os/desktop:dev`
> 
> Please note that this is still a test build and **NOT** ready for production use.

Incredible work!
Is this a generic arm64 build or made for specific devices?
Is it possible to make a build working for Apple Silicone or Qualcomm Snapdragon X Elite?

**Comment by NN708** [comment]
> Is this a generic arm64 build or made for specific devices?

This is a generic ARM64 build.

> Is it possible to make a build working for Apple Silicone or Qualcomm Snapdragon X Elite?

Whether it works on specific hardware depends on the hardware's level of mainline support. For example, on Apple Silicon, you'll need to build a custom image that includes packages like `linux-asahi` and `mesa-asahi` from the [Debian Bananas Team](https://wiki.debian.org/Teams/Bananas).

**Comment by JeremiahCornelius** [comment]
I just found this, and hope I haven't duplicated too much effort.

I'm doing something kind of crazy, building a VanillaOS Installer ISO for ARM - aarch64. I hacked together a build harness for reproducibility in bash.

Arm laptops are weird. As SoC devices, and terrible non-Windows ACPI support, these need a dtb file like a single-board computer, and often a purpose-built kernel and custom firmware. My harness does this work with debs I built and borrowed  for Ubuntu Concept ARM.

I'm able to get arm binaries for VIB and pull an arm Pico container - and this builds a bootable ISO that starts on my system both with Orchid v2 and with v3 from github.

HOWEVER:
The squashfs is consistently incomplete for running vanilla-installer. I get a Plymouth boot splash, then dump to a command prompt. Running vailla-installer from the prompt complaains mutter isn't present. In fact, no Gnome binaries are in the booted system.

Has anyone gotten further with this? Has someone with success here got pointers and advice?

I think if I get further here with these existing efforts and my harness, we have a way to supply the right kernels and firmware dumps to support devices known by Ubuntu Concept to run VanillaOS Orchid - like my experimental HP Omnibook 5 16".

QUESTION:
What is the complete workflow to reproduce the build of your 2.2GB ISO? Do you use an amd64 build host? The method in this git relies on a pico image that is unavailable on my arm64 build host.


Thanks!

**Comment by NN708** [comment]
> What is the complete workflow to reproduce the build of your 2.2GB ISO? Do you use an amd64 build host? The method in this git relies on a pico image that is unavailable on my arm64 build host.

The installer ISO for ARM64 has not yet been released. To build it, you will need to use the `ghcr.io/vanilla-os/pico:dev` image.

**Comment by JeremiahCornelius** [comment]
> > What is the complete workflow to reproduce the build of your 2.2GB ISO? Do you use an amd64 build host? The method in this git relies on a pico image that is unavailable on my arm64 build host.
> 
> The installer ISO for ARM64 has not yet been released. To build it, you will need to use the `ghcr.io/vanilla-os/pico:dev` image.

Great. My build harness now successfully produces an aarch64 bootable VanillaOS Orchid image, with a kernel supplied for Qualcom Snapdragon X, appropriate firmware objects, and a dtb that provides device control for trackpad and full USB ports.

The harness is designed so that any custom kernel debs, dtb, and firmware can be supplied besides the defaults for hp omnibook-5.

I can invoke 'IGNORE_CPU=1 vanilla-installer', and specify ghcr.io/vanilla-os/desktop:dev for the installer image. The result on the target does not preserve the image modifications that produced the installation ISO, so the (rapidly) installed target system enters a boot loop from the ABroot prompt after restart.

I'm open to suggestions for getting past this next obstacle. The obvious one is to privately host the custom OCI on a registry. This might be onerous for publishing the harness for use by others.

QUESTION: Is core-image the only one I need to supply? Pointers here are welcome. 

Also, would it be possible to include the required OCI image(s) on the installation ISO itself? Can vanilla-installer accept filesystem locations, as it does for container registries?

Thanks very much.

**Comment by NN708** [comment]
> I'm open to suggestions for getting past this next obstacle. The obvious one is to privately host the custom OCI on a registry. This might be onerous for publishing the harness for use by others.

You can just use GHCR to host it.

> QUESTION: Is core-image the only one I need to supply? Pointers here are welcome.

Actually, it's [Vanilla-OS/desktop-image](http://github.com/Vanilla-OS/desktop-image).

> Also, would it be possible to include the required OCI image(s) on the installation ISO itself? Can vanilla-installer accept filesystem locations, as it does for container registries?

Not currently, but contributions are welcome.

**Comment by JeremiahCornelius** [comment]
> > Also, would it be possible to include the required OCI image(s) on the installation ISO itself? Can vanilla-installer accept filesystem locations, as it does for container registries?
> 
> Not currently, but contributions are welcome.

Good news, I have a build harness written in bash, that generates ISO Vanilla Reunion installation ISO images using VIB for ARM aarch64 laptops with Snapdragon X SoC and UEFI firmware.
User-supplied kernels, device trees and firmware mods are integrated from deterministic locations in the build tree. Special kernel command arguments are preserved in the ISO and correctly configured in resulting target laptop devices.

Unique to builds by this harness, the image build for Desktop/Gnome OCI is by default, hosted on the ISO image itself. Included at boot-time is a minimal unsecured OCI image repository limited to localhost. This allows the entire installation to complete from a single, ~ 6GB iso without pulling from a remote docker-style repository, enabling rapid build/debug cycles.

Currently, the Snapdragon X Vanilla OS build harness is a late-alpha stage, but I have two successful candidates with repeated, deterministic installations — a Microsoft Surface 7 13" Snapdragon X Elite and an HP Omnibook 5 16" Snapdragon X Plus. Both with standard I/O peripherals and Adreno graphics acceleration, and reasonable mobile battery/power management. By core design for building, other models are easily supported, given the right user supplied kernel, etc. With minor adaptation, this could be the basis to build a Vanilla OS install image for Apple M1/M2 targets, including Mac Mini and Macbook Air models.

Snapdragon X is a rapidly moving target, with most support being done by the Ubuntu Snapdragon Concept project and a couple of leading edge contributors to those efforts, without whom this wouldn't be possible.

I'm six weeks into work on this, and will be sharing a github repository for the effort after hitting a couple more milestones.
